pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,404 @@
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	// This is guaranteed to avoid conflicting with the main routing table ID
+	// (253-255) because the maximum number of ENI devices on a node is 15 (see
+	// pkg/aws/eni/limits.go). Because the interface number is monotonically
+	// increasing and the lowest available number is reused when devices are
+	// added / removed. This means that the max possible table ID is 25.
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// onUpgradeInstallRuleAndRoutes migrates the ENI datapath for the upgrade
+// case. It accepts the old table ID for the existing rules to migrate. `new`
+// is the new table ID that the rules and routes should be referring to.
+func onUpgradeInstallRuleAndRoutes(
+	rule netlink.Rule,
+	routes []netlink.Route,
+	oldTable, newTable int,
+) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	revert, err := copyRoutes(routes, oldTable, newTable)
+	revertStack.Extend(revert)
+	if err != nil {
+		return revertStack, err
+	}
+
+	revert, err = createNewRule(
+		rule,
+		linux_defaults.RulePriorityEgress,
+		oldTable,
+		linux_defaults.RulePriorityEgressv2,
+		newTable,
+	)
+	revertStack.Extend(revert)
+	if err != nil {
+		return revertStack, err
+	}
+
+	return revertStack, nil
+}
+
+// onUpgradeDeleteRuleAndRoutes removes the old ENI datapath for the upgrade
+// case. The rule to be removed as well as the routes are passed in.
+func onUpgradeDeleteRuleAndRoutes(
+	rule netlink.Rule,
+	routes []netlink.Route,
+	oldTable int,
+) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	revert, err := deleteOldRule(rule)
+	revertStack.Extend(revert)
+	if err != nil {
+		return revertStack, err
+	}
+
+	revert, err = deleteOldRoutes(routes)
+	revertStack.Extend(revert)
+	if err != nil {
+		return revertStack, err
+	}
+
+	return revertStack, nil
+}
+
+func copyRoutes(routes []netlink.Route, from, to int) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	// In ENI mode, we only expect two rules:
+	//   1) Link scoped route with a gateway IP
+	//   2) Default route via gateway IP
+	// We need to add the link-local scope route to the gateway first, then
+	// routes that depend on that as a next-hop later.  If we didn't do this,
+	// then the kernel would complain with "Error: Nexthop has invalid
+	// gateway." with an errno of ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := netlink.RouteReplace(&route); err != nil {
+				return revertStack, fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			revertStack.Push(func() error {
+				if err := netlink.RouteDel(&route); err != nil {
```

Comment:
I suspect we're in "good enough" territory here, but note that "RouteReplace" may overwrite an existing route, which means that the corresponding revert operation would be to reinstate the replaced route, not to delete the new one.

That said, I think we're probably going a bit too far off the beaten track considering that we assume Cilium is the one administering these route tables?

Reply:
(Unless we're specifically worried about the case where the ifindex and the new route table are the same number..?)